PP-326, PP-327, PP-328, PP-331, PP-332, PP-333, PP-335: Initial checkin of hook to support cgroup functionality

JonShelley opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Enter your JIRA issue id below -->
<!--- If a JIRA issue is not available, please first create one at pbspro.atlassian.net -->
* PP-326, PP-327, PP-328, PP-331, PP-332, PP-333, PP-335

#### Problem description
* Users are able to start jobs and consume more resources then requested. This enables the user to adversely impact other users as well as system.

#### Cause / Analysis
* Jobs are not place in containers that limit their resource usage to what was requested

#### Solution description
* Linux cgroup capability allows the admin to restrict the job to the resources requested. This enables jobs to share nodes and ensure that each job has access to the resources requested. It also enables the admin to get accurate resource usage without having to poll for it.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD/PP-325+Support+Cgroups)**.
- [x] I have added tests to cover my changes. To create automated tests, see **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**.
- [x] All new and existing automated tests have passed. See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**.
- [x] I have attached automated (PTL)/manual test logs to the associated Jira ticket.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

mike0042 commented:
Closing pull request to kick off travis-ci check.

mike0042 commented:
Reopening pull request to run checks.

bayucan commented:
I've looked through all the files. The PTL tests part seemed fine. My comment is with the pbs_groups.py contents:

The following lines should be 'pbs.event().accept()":

330	      pbs.accept()

715		pbs.accept()


953		pbs.accept()

--

In the following lines:

# For CentOS 7 we need to remove a *:* rwm from devices.list we 
2071             # before can add anything to devices.allow. Otherwise our changes 
2072             # are ignored. Check to see if a *:* rwm is in devices.list. 
2073             # If so remove it 
2074             if "a *:* rwm\n" in devices_allowed: 
2075                 value = "a *:* rwm" 
2076                 self.write_value(os.path.join(path, 'devices.deny'), value) 
2077             else: 
2078                 # Verify that the following devices are not in devices.list 
2079                 pbs.logmsg(pbs.EVENT_DEBUG3, 
2080                            "Removing access to the following: %s" % 
2081                            devices_list) 
2082             for device_str in devices_list: 
2083                 value = "c %s rwm" % device_str 
2084                 self.write_value(os.path.join(path, 'devices.deny'), value) 
-> shouldn't lines 2082-2084 be indented to fall under the 'else:' case? Or
maybe take out the 'else:' part altogether since the devices.deny of
devices_list will happen in both cases.

---
In the following lines:

             if vmem_enabled: 
2975                 if mem_requested is None: 
2976                     pbs.logmsg(pbs.EVENT_DEBUG, 
2977                                ("%s: vmem not requested, " + 
2978                                 "assigning %s to cgroup") % 
2979                                (caller_name(), vmem_limit)) 
2980                 hostresc['vmem'] = pbs.size(vmem_limit) 

-> Shouldn't this be 'vmem_requested' rather than 'mem_requested'? 

---

In the following lines, I think the fd.close() is unnecessary since we execute via 'with' which automatically closes the close when done:

3102	 with open(tasks_file, 'r') as fd: 
3103             for line in fd: 
3104                 os.kill(int(line.strip()), signal.SIGKILL) 
3105         fd.close() 


--

Wow, didn't one can do a time.sleep(.25)! I thought it was only in full seconds
increments.


--

In the following function, shouldn't we close the opened file? Or perhaps
use the 'with' construct. Same deal with all the other __get_* functions.

     def __get_mem_failcnt(self, path): 
3321         try: 
3322             return int(open(os.path.join(path, 
3323                        "memory.failcnt"), 'r').read().strip()) 
3324         except: 
3325             return None

mike0042 commented on src/hooks/cgroups/pbs_cgroups.json:
```diff
@@ -0,0 +1,54 @@
+                "memory":
+                {
+                        "enabled"               : true,
+                        "default"               : "250MB",
```
The default should be 256MB, as it is for memsw.

JonShelley commented:
Al,

Thanks for the review. I have corrected the pbs.event().accept() lines.

I have removed the else statement on lne #2078

I have removed the close file descriptor at line #3105

As for the comment about closing the file in the __get_* functions, I believe that once you leave the scope of the function that it is gets cleaned up automatically. If you still feel like I need to explicitly close it please let me know and I will make the changes

JonShelley commented:
Mike,

I will update the mem value in the config to 256MB.

mike0042 commented:
I see that eatmem.c is referenced by the tests, but I don't see actual file in the pull request. I have a small python script that does something very similar to what eatmem.c does. It takes two arguments... the number of chunks and the chunk size. Running "eatmem.py 72 8" will allocate 72 chunks of 8MB (576MB total). It should have a .py extension, or no extension at all, but JIRA won't let me attach files like that.
[eatmem.txt](https://github.com/PBSPro/pbspro/files/414126/eatmem.txt)

mike0042 commented:
Filed PP-404 to address comments from this pull request after checkin.
